### PR TITLE
Fix on command makecolumns

### LINKED ISSUE #1900: Alignment problem when using \makecolumns
The right edge of the frames which are bond to the content one doesn't shift as the number of columns increases.

notaligned

So I did the following changes:

```lua
  self:registerCommand("makecolumns", function (options, _)
    local aligned = {}
    options.columns = options.columns or 2
    for key in pairs(SILE.frames) do
      -- this must be checked before makecolumns() call, since the function changes the right() value of content
      if SILE.getFrame(key):right() == SILE.getFrame("content"):right() and key ~= "content" then
         table.insert(aligned, key)
      end
    end

    makecolumns(options)

    for _, frame in ipairs(aligned) do
      SILE.getFrame(frame):constrain("right", SILE.getFrame("content_col"..options.columns-1):right())
    end
  end, "Split the current frame into multiple columns")
```
And now they do:

aligned

---

The footnote frame is also overwritten if its top and the bottom of content frame are linked, I know this has more to do with the frameset definition than with the makecolumns method, but I think it worth mentioning...

footnote

```sil
\begin[papersize=a5]{document}
\script[src=packages/frametricks]
\script[src=packages/lorem]
\script[src=packages/footnotes]
\makecolumns
\lorem[words=100]
\showframe[id=all]
\footnote{\lorem[words=20]}
\lorem[words=500]
\end{document}
```

But when the bottom of content is set to a constant value, e.g. 80%ph like this, it works well:

footnote2

### PR DESCRIPTION
**Pull request by jodros**
Fixes #1900

### CODE REVIEW

**Comment by alerque** [comment]
c.f. #420

**Review by alerque — COMMENTED** [review]
How does this fare when *reducing* the number of columns (as noted is a problem in #248) and/or subsequently changing the value after doing it once? Fixing those cases might not be blockers for this PR but we should at least track whether they are fixed or not.

**Comment by jodros** [comment]
I began to play with \makecolumns just after reading the issue #248, so I have interest in finding a solution for it as well...

> c.f. #420

You mention have done a "dirty hack", well, this description:

> adding ALL the broken frame id's to the insertion steal from list and iterating over ones that exist on a given page

looks like my own approach, why would it be a "dirty hack"? How should a clean fix look like?

**Comment by alerque** [comment]
Sorry I seem to have inadvertently edited *your comment* instead of *reply to selection*! I restored your comment and here is my reply:

> why would it be a "dirty hack"? How should a clean fix look like?

It doesn't take into account things like the possibility of splitting frames (say to add a float) after starting columns, or ending columns after they are started. You don't want insertions stealing space from anything but the actually *adjacent* frames. Ones that were adjacent at one time but later got split or whatever shouldn't be eligible for theft.

**Comment by alerque** [comment]
LOL, so I tried to fix this to work in any direction, but it turns out the column splitter assumes the direction anyway, and this isn't the right output either:

20231027_14h39m11s_grim

Good times. I do think the code I added will be useful later when `makeoptions()` takes into account the frame advance direction, but for now I'll backtrack a little bit to assuming "right" again.

**Comment by alerque** [comment]
@jodros How does my last commit here look to you? It turns out it isn't too useful since it still only works in the `:right()` direction, but I hopefully this will make it easier not harder to eventually expand the column tooling to automatically generate columns in any advance direction.

File: tests/makecolumns.sil [hunk 1]
```diff
@@ -0,0 +1,5 @@
+\begin[papersize=a5]{document}
+\script[src=packages/frametricks]
```
Comment (alerque) [inline tests/makecolumns.sil:2]: @jodros I'll fix this test, but just for your own reference in the future, the non-deprecated syntax for this is:

```suggestion
\script[src=packages/frametricks]
\use[module=packages.frametricks]
```

**Review by alerque — CHANGES_REQUESTED** [review]

**Review by alerque — DISMISSED** [review]

**Comment by alerque** [comment]
Please don't merge from master into feature branches unless there is a real reason to do so, such as the feature being worked on being dependent on something else that landed. It just makes history tracking / git blame / etc. much harder to work with.

**Comment by Omikhleia** [comment]
> Please don't merge from master into feature branches unless there is a real reason to do so, such as the feature being worked on being dependent on something else that landed. It just makes history tracking / git blame / etc. much harder to work with.

Rebase is _always_ a good friend ;)

**Comment by jodros** [comment]
> Please don't merge from master into feature branches unless there is a real reason to do so, such as the feature being worked on being dependent on something else that landed. It just makes history tracking / git blame / etc. much harder to work with.

Sorry, I didn't know it was a problem, I just reverted the merging right now.

**Review by alerque — APPROVED** [review]

**Pull request merged**